### ISSUE TITLE
[discussion] Declaring typedefs on classes with memory management strategies

### ISSUE THREAD

**Issue by bmcdonald3** [chapel-lang/chapel#20082]
## Summary

In the code snippet below, a `type myType = shared MyClass;` has been declared and so all uses of `myType` will be shared. Since shared objects can also be borrowed, being able to add a `borrowed` to the `myType` instance would be valuable for cases when you would like to coerce the type to `borrowed`, like in the `+` operator below. 
```
class MyClass {
  var a: int;
}

type myType = shared MyClass;

inline operator +(l: borrowed myType, r: borrowed myType): myType {
  return new myType(l.a + r.a);
}

var val1 = new myType(5);
var val2 = new myType(10);

var val3 = val1 + val2;
```
Here, the code will not compile, because it is expanded to `borrowed shared MyClass`, but it could make sense for that borrowed to "overload" that shared, since we want to borrow from a shared class. As it works today, the single typedef is not particularly valuable because if you want to change from using `shared`, you have to write out the full class or cast or something, rather than just specifying in front of the type.

## Proposal

If `borrowed` were able to be overloaded onto `shared` class instances, this would make the typedef more useful and avoid needing the second typedef.

What is being proposed here is a slight change to Chapel's conception of `borrowed` as a memory management strategy, where it could instead be thought of as a memory management strategy modifier on top of owned/shared, rather than only its own thing.

## Workaround

A workaround that has been suggested for the above code is to add a `type myBType = myType: borrowed;` and that achieves the desired result, but seems unfortunate that you would need those two separate typedefs.

**Comment by bradcray** [comment]
FWIW, my first reaction to this question is "Why is an operator being defined on a type alias rather than on the type itself (by the type author, say)."  I.e., if `MyClass` was a natural candidate for a `+` operator, I'd expect `+` to be defined on it rather than on a type alias for a specific flavor of it.

That said, checking back to the gitter question that spawned this topic — https://gitter.im/chapel-lang/chapel?at=62b47155ef5ee44882c14b3a — in that context, it was a standalone function, where it makes much more sense to choose to want to take a particular flavor of a class instead.  So to others, I'd say "Don't necessarily get sidetracked by the operator in the OP here like I did."

**Comment by bradcray** [comment]
> Would it make sense to allow myType.borrow() calls as a way of specifying "I know this type alias includes a memory management strategy that supports borrowing from it, require that this is a borrow of that type"?

That seems virutally identical to what `myType: borrow` does today, so gets into the standard territory of "should we have two ways of doing things?" question for me.  Note that we don't distinguish in any way between "borrowed from shared", "borrowed from owned" or "borrowed from unmanaged", so can't (without additional changes to the language) distinguish between the source of the borrow.

I was disappointed to find that I couldn't find a way to write `myType:borrowed` as a formal type, which seemed surprising.

**Comment by mppf** [comment]
> I was disappointed to find that I couldn't find a way to write `myType:borrowed` as a formal type, which seemed surprising.

Yes, I would expect `inline operator +(l: (myType:borrowed), r: (myType:borrowed)): myType` to work, but it does not.

One thing that does work is

``` chapel
proc toBorrow(type t) type {
  return t:borrowed;
}
inline operator +(l: toBorrow(myType), r: toBorrow(myType)): myType { }
```

**Comment by mppf** [comment]
Also I don’t think it’s necessarily the case that we can’t support `borrowed shared MyClass`. At some point in the past we had a “the outermost one wins” interpretation of this sort of thing but then we decided that was confusing and changed it into an error. IIRC that confusion was also connected to how `new` behaves. However I would have to do some archaeology to find the justification from the time.

**Comment by bradcray** [comment]
> At some point in the past we had a “the outermost one wins” interpretation of this sort of thing but then we decided that was confusing and changed it into an error.

I remember this discussion, and was definitely on the side of finding it a bit confusing.  What I can't recall was whether it was limited to `borrowed x MyClass` or would also apply to things like `owned shared Myclass` ("owned wins").  I think one downside of `borrowed shared MyClass` is that it might suggest that it's a different type from `borrowed owned MyClass`—that is, "I'm a borrow of a shared, whereas you're a borrow of an owned."  Again, if I'm keeping things straight, I don't think we distinguish the source of a borrow once it's created (at compile time or execution time).

**Comment by mppf** [comment]
One thing that wasn't clear to me until today is that the goal of 
`type myType = shared MyClass;` (vs writing `shared MyClass` instead of `myType` in function signatures etc) is to make it easier to switch between `owned` and `shared` without modifying all of the places where `myType` is used.

**Comment by vasslitvinov** [comment]
It makes less sense to allow literally `borrowed shared MyClass`. Allowing `borrowed myType` is more reasonable.

However, we could just direct users to use `(myType: borrowed)` instead of `borrowed myType`, at the expense of a couple of extra symbols.

I think Michael's `operator +(l: (myType:borrowed), r: (myType:borrowed)): myType` should work. Why would we not want to allow that?

**Comment by bradcray** [comment]
> I think Michael's operator +(l: (myType:borrowed), r: (myType:borrowed)): myType should work. Why would we not want to allow that?

I agree that we want to allow that, and don't think anyone has argued otherwise (but nobody has signed up to fix it yet either).  I suspect the problem is that our parser doesn't support general expressions in the formal argument type position.

That said, even if we did support it, it's not obvious whether it would satisfy the OP.

(Also, I want to claim credit where due.  This was my proposal in https://github.com/chapel-lang/chapel/issues/20082#issuecomment-1167613984   :D )

**Comment by jabraham17** [comment]
@bmcdonald3 can this be closed? We have come to consensus that our intent is to change types via casts like `myOwnedType:borrowed`, and #23109 allows this syntax in limited form for function formals.

**Comment by bmcdonald3** [comment]
> We have come to consensus that our intent is to change types via casts like myOwnedType:borrowed

@jabraham17 Is what you are saying is that, based on these decisions, we are going to support something like this:

```chpl
type t = shared myType;
proc myFunc(arg: borrowed t) { ... }
```

If that is the case, then I think this issue can be closed, but it might make sense to close it when the actual PR is filed instead of when the decision has been made.

**Comment by jabraham17** [comment]
No, the intention is that we will support the following
```chapel
type t = shared myType;
proc myFunc(arg: t:borrowed) { ... }
```

**Comment by bmcdonald3** [comment]
OK, that looks good to me. I think we can consider this resolved as a part of that, but I still would probably close it when the actual implementation is done to have it linked with the PR, but if there is a particular reason to close this now, I wouldn't put up a fight.

**Comment by jabraham17** [comment]
I have merged the partial fix for `myType:borrowed` and have spun out a separate issue (#22994) to track the specific feature.
